pkg/controller/logstorage/linseed/linseed_controller.go

```diff
@@ -0,0 +1,389 @@
+			return fmt.Errorf("log-storage-controller failed to watch Tenant resource: %w", err)
+		}
+	}
+
+	// The namespace(s) we need to monitor depend upon what tenancy mode we're running in.
+	// For single-tenant, everything is installed in the tigera-manager namespace.
+	installNS := render.ElasticsearchNamespace
+	truthNS := common.OperatorNamespace()
+	if opts.MultiTenant {
+		// For multi-tenant, we could be installed in any number of namespaces.
+		// So, watch the resources we care about across all namespaces.
+		installNS = ""
+		truthNS = ""
+	}
+
+	// Watch secrets this controller cares about.
+	secretsToWatch := []string{
+		render.TigeraElasticsearchGatewaySecret,
+		render.TigeraLinseedSecret,
+		render.LinseedTokenSecret,
+		monitor.PrometheusClientTLSSecretName,
+	}
+	for _, ns := range []string{truthNS, installNS} {
+		for _, name := range secretsToWatch {
+			if err := utils.AddSecretsWatch(c, name, ns); err != nil {
+				return fmt.Errorf("log-storage-controller failed to watch Secret: %w", err)
+			}
+		}
+	}
+
+	// Catch if something modifies the resources that this controller consumes.
+	if err := utils.AddServiceWatch(c, render.ElasticsearchServiceName, installNS); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch the Service resource: %w", err)
+	}
+
+	// Check if something modifies resources this controller creates.
+	// TODO
+	if err := utils.AddServiceWatch(c, esgateway.ServiceName, installNS); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch Service resource: %w", err)
+	}
+	if err := utils.AddServiceWatch(c, render.LinseedServiceName, installNS); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch Service resource: %w", err)
+	}
+	if err := utils.AddDeploymentWatch(c, render.LinseedServiceName, installNS); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch Deployment resource: %w", err)
+	}
+	if err := utils.AddDeploymentWatch(c, esgateway.DeploymentName, installNS); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch the Service resource: %w", err)
+	}
+
+	return nil
+}
+
+func (r *LinseedSubController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling LogStorage - Linseed")
+
+	// When running in multi-tenant mode, we need to install Linseed in tenant Namespaces. However, the LogStorage
+	// resource is still cluster-scoped (since ES is a cluster-wide resource), so we need to look elsewhere to determine
+	// which tenant namespaces require a Linseed instance. We use the tenant API to determine the set of namespaces that should have a Linseed.
+	var tenant *operatorv1.Tenant
+	var err error
+	if r.multiTenant {
+		if request.Namespace == "" {
+			// In multi-tenant mode, we only handle namespaced reconcile triggers.
```

Comment:
With the introduction of the EnqueueAllTenants EventHandler, I might be able to get rid of this branch actually. Need to look into that.